### PR TITLE
[Fix] Only display sources with an icon

### PR DESCRIPTION
**Pull request by leb2dg**

### CODE REVIEW

File: api/views/workflow.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ class SourceViewSet(viewsets.ReadOnlyModelViewSet):
     serializer_class = SourceSerializer
 
     def get_queryset(self):
-        queryset = Source.objects.all()
+        queryset = Source.objects.exclude(icon='')
```
Comment (chrisseto) [inline api/views/workflow.py]: Should this be `icon__isnull=True`?

File: share/graphql/query.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def resolve_me(self, args, context, info):
     @graphene.resolve_only_args
     def resolve_sources(self, limit=25, offset=0):
         # TODO Long title will need an index
-        return models.ShareUser.objects.exclude(long_title='').exclude(favicon='').order_by('long_title')[offset:offset + limit]
+        return models.ShareUser.objects.exclude(long_title='').exclude(icon='').order_by('long_title')[offset:offset + limit]
```
Comment (chrisseto) [inline share/graphql/query.py]: See above comment about icon.

Does this result in an `AND` or an `OR` query/ What is it intended to do?

File: bots/elasticsearch/tasks.py [hunk 3]
```diff
@@ -117,8 +117,13 @@ def do_run(self, es_url=None, es_index=None):
     def bulk_stream(self, es_index):
         Source = apps.get_model('share.Source')
         opts = {'_index': es_index, '_type': 'sources'}
-        for source in Source.objects.exclude(long_title='').all():
-            yield {'_op_type': 'index', '_id': source.name, **self.serialize(source), **opts}
+
+        for source in Source.objects.all():
+            # remove sources from search that don't appear on the sources page
+            if source.icon == '':
```
Comment (chrisseto) [inline bots/elasticsearch/tasks.py]: `if source.icon_id is None:` would be preferable here.

`source.icon` will load the entire favicon from the database every time.

File: share/graphql/query.py (hunk 2 above)
Comment (leb2dg) [inline share/graphql/query.py]: It's an `OR` query (pretty sure). It applies one exclude then the other https://docs.djangoproject.com/en/dev/topics/db/queries/#chaining-filters. Would it be better to use a Q object https://docs.djangoproject.com/en/dev/topics/db/queries/#complex-lookups-with-q-objects?

Comment (chrisseto) [inline share/graphql/query.py]: Looks like `long_title` doesn't exist any more and chaining seems to result in an `AND`

```python
In [6]: str(Source.objects.exclude(long_title='').exclude(icon='').query)
Out[6]: 'SELECT "share_source"."id", "share_source"."name", "share_source"."long_title", "share_source"."home_page", "share_source"."icon", "share_source"."user_id" FROM "share_source" WHERE (NOT ("share_source"."long_title" = ) AND NOT ("share_source"."icon" =  AND "share_source"."icon" IS NOT NULL))'
```

File: api/views/workflow.py (hunk 1 above)
Comment (leb2dg) [inline api/views/workflow.py]: When clearing an image file through the admin interface, the field isn't nulled. I can't find any documentation on this though?

File: share/graphql/query.py (hunk 2 above)
Comment (aaxelb) [inline share/graphql/query.py]: `NOT (a OR b) <-> NOT a AND NOT b`

Comment (aaxelb) [inline share/graphql/query.py]: And this should be showing Sources, not ShareUsers

File: api/views/workflow.py (hunk 1 above)
Comment (aaxelb) [inline api/views/workflow.py]: Might be better to make it just `blank=True`, not `null=True`, and keep this exclude the same

File: bots/elasticsearch/tasks.py (hunk 3 above)
Comment (aaxelb) [inline bots/elasticsearch/tasks.py]: No, just `if source.icon:` will do. The field returns a little `FieldFile` instance, not the actual favicon.

File: share/graphql/query.py (hunk 2 above)
Comment (chrisseto) [inline share/graphql/query.py]:
```python
In [2]: str(Source.objects.filter(long_title='foo').exclude(icon='').query)
Out[2]: 'SELECT "share_source"."id", "share_source"."name", "share_source"."long_title", "share_source"."home_page", "share_source"."icon", "share_source"."user_id" FROM "share_source" WHERE ("share_source"."long_title" = foo AND NOT ("share_source"."icon" =  AND "share_source"."icon" IS NOT NULL))'
```

```python
In [3]: str(Source.objects.filter(long_title='foo').filter(icon='1').query)
Out[3]: 'SELECT "share_source"."id", "share_source"."name", "share_source"."long_title", "share_source"."home_page", "share_source"."icon", "share_source"."user_id" FROM "share_source" WHERE ("share_source"."long_title" = foo AND "share_source"."icon" = 1)'
```

File: bots/elasticsearch/tasks.py (hunk 3 above)
Comment (chrisseto) [inline bots/elasticsearch/tasks.py]: Funky, I stand corrected. 👍

File: share/graphql/query.py (hunk 2 above)
Comment (leb2dg) [inline share/graphql/query.py]: Ah sorry negative query confusion. I meant ~(P OR Q) which is the same as ~P AND ~Q. Did not know how to see the query though, that's neat.

I didn't look into that method though...seems like the whole thing needs to change if it's being used?

Comment (leb2dg) [inline share/graphql/query.py]: Ah Abram beat me

File: share/graphql/query.py [hunk 4]
```diff
@@ -41,7 +41,7 @@ def resolve_me(self, args, context, info):
     @graphene.resolve_only_args
     def resolve_sources(self, limit=25, offset=0):
         # TODO Long title will need an index
```
Comment (aaxelb) [inline share/graphql/query.py]: While you're in here, this TODO can go away, long_title is uniquely indexed now

File: api/views/workflow.py (hunk 1 above)
Comment (chrisseto) [inline api/views/workflow.py]: It looks like django will turn this into the correct query? But not with a `.filter`, odd.

```python
In [9]: str(Source.objects.filter(icon='').query)
Out[9]: 'SELECT "share_source"."id", "share_source"."name", "share_source"."long_title", "share_source"."home_page", "share_source"."icon", "share_source"."user_id" FROM "share_source" WHERE "share_source"."icon" = '

In [10]: str(Source.objects.exclude(icon='').query)
Out[10]: 'SELECT "share_source"."id", "share_source"."name", "share_source"."long_title", "share_source"."home_page", "share_source"."icon", "share_source"."user_id" FROM "share_source" WHERE NOT ("share_source"."icon" =  AND "share_source"."icon" IS NOT NULL)'
```

Actually looked at the DB, thought that icon was going to be FK not a string.

@aaxelb's suggest is most correct, I think.

Comment (aaxelb) [inline api/views/workflow.py]: File/ImageFields are weird, and I feel like I abused them pretty hard to store icons in the DB. The string would usually be a path on a filesystem...

Comment (chrisseto) [inline api/views/workflow.py]: You can blame me for that one. 
#TODO

**Comment by chrisseto** [comment]
Will merge once tests pass.

**Comment by leb2dg** [comment]
A 404 is returned when trying to look at a record pushed by a trusted user when that user doesn't have an icon. Looking into it a bit more.

**Comment by leb2dg** [comment]
@chrisseto fixed the 404 thing. Now, if no icon, it looks like:
screen shot 2017-03-29 at 4 12 54 pm

**Comment by chrisseto** [comment]
Wunderbar! waiting on tests

**Pull request merged**